docs/spec/connection-handshake/L1_2.md

```diff
@@ -0,0 +1,264 @@
+# IBC Connection Handshake Spec
+
+> Work in progress.
+
+> All the content until section [2.2 Protocol] is stable and ready for review.
+
+> Reviews are welcome.
+
+
+## L1. Problem statement / outside view
+
+Suppose there are two chains (e.g., distributed ledgers) that execute independently of each other.
+What should be a basic abstraction for representing a path of communication between the two?
+We use the term _connection_ to represent this path of communication.
+The purpose of the IBC connection handshake protocol, at the highest level of abstraction, is to enable two chains to establish a connection with each other [[ICS 003](#references)].
+Throughout the rest of this document, we will often use the abbreviation "CH" to stand for "connection handshake" problem (or protocol, depending on the context).
+
+#### Sequential problem statement
+
+Instead of _chains_, at this abstract level of specification we can reason in terms of _processes_.
+There are two processes in our problem statement, which we can label explicitly as an _initiator_ (or source) process and a _responder_ (or destination) process.
+
+
+###### Problem statement and guarantees
+A connection handshake allows, briefly, that two processes agree on a consistent view of their state, and each process allocates a new connection associated with this state.
+We identify the following requirements that the connection handshake should guarantee:
+
+- **[CH/Seq/1/Agreement]** If the initiator finishes the handshake, establishing a new connection `ci` with the responder, and the responder finishes the handshake obtaining the connection `cr`, then `ci == cr`.
+In other words, the two ends of the connection (one at the initiator and the other at the responder) are consistent.
+
+
+- **[CH/Seq/2/Termination]** Eventually, the two processes should both finish the handshake, establishing a connection.
+
+Requirement **[CH/Seq/1/Agreement]** represents the basic safety guarantee, while **[CH/Seq/2/Termination]** is the basic liveness guarantee we would like to have.
+
+The diagram below sketches the high-level problem statement.
+
+```
+                        IBC Connection
+     +-----------+    Handshake Protocol    +-----------+
+     |           |<-------- step 1 -------->|           |
+     | Initiator |    (execute handshake)   | Responder |
+     |           |                          |           |
+     |          _| (connection established) |_          |
+     |         |c|<------------------------>|c|         |
+     |          ▔|                          |▔          |
+     +-----------+                          +-----------+
+```
+
+
+## L2. Protocol specification / protocol view
+
+In this section we attempt to make the problem specification more concrete.
+We define what the processes are, their fault-model, as well as communication and synchrony assumptions.
+
+### 2.1 System model specification
+
+#### System model
+
+A **chain** is a state machine.
```

Comment:
```suggestion
A **chain** is a deterministic state machine.
```

Reply:
Hmm, I wasn't aware we require chains to be deterministic -- I was thinking that maybe randomized-based algorithms, e.g., Bitcoin, are allowed in IBC? But I guess when viewed from the "outside" (from the perspective of other chains and IBC) each chain must appear to behave deterministically -- is that what you mean?

Reply:
Bitcoin has randomness in its consensus algorithm, but the state machine it replicates is still a deterministic state machine (deterministic function of the txs).